### ISSUE TITLE
Dismantle dependency installation and setup. 

### ISSUE THREAD

**Issue by batchor** [NodeBB/NodeBB#12182]
### Description

The current `nodebb setup` step includes both dependency installation and setup. I couldn't find a good way for docker compose deployment.

In my case I want to separate `nodebb` and `mongodb` into two containers. However, in the image building stage where all the dependencies are installed, the `nodebb setup` needs to connect to mongodb server, which is not ready yet.

I think dismantling dependency installation and setup (mainly mongodb connection) would be better for docker deployment. Dependencies are installed during building the docker image, and mongodb connection check is performed inside containers.

**Comment by oplik0** [comment]
How are you installing dependencies before setup? Are you sure you put `package.json` in NodeBB folder when doing the setup?

Because NodeBB needs some dependencies installed during setup, but it only does install then if they're not installed correctly already.
The process is basically:
1. check if package.json exists
2. check if default packages (ones that are also listed in `install/package.json`) satisfy their semver strings
3. check that versions of a couple of packages (I think mostly the ones used in CLI directly) are up to date

If any of these fails, NodeBB will install dependencies. Otherwise it won't bother outside of explicitly using `upgrade` command.

The way to do manual installation is fairly trivial - copy `install/package.json` to just `package.json` and run `npm install`

For example, the dockerfile in `develop` branch runs npm install at startup to ensure things are up to date (npm not doing anything is surprisingly fairly fast, it's actually faster than pnpm and yarn when it already has everything installed), but otherwise dependencies are installed during build (and there is some additional complexity related to supporting multiple architectures because it turns out npm is very slow on emulated arm VMs and it's much faster to just rebuild native dependencies afterwards)

**Comment by batchfly** [comment]
@oplik0 Oh I couldn't find `package.json` before. Now I am able to install dependencies with `install/package.json` during docker image building.

Just a little bit curious why not put `package.json` in the root directory which is a convention in JavaScript projects.

**Comment by oplik0** [comment]
Because NodeBB is abusing git for update management :)

I'm joking about the abuse, but the issue is that since git is being used to update the codebase, you don't want the user to change any tracked files. Otherwise trying to update from git will result in either merge conflicts (if using pull) or loss of the user changes (if using reset). 

Since `package.json` changes due to plugin installation, and NodeBB has upgrade tools regardless (for e.g. db migrations), the solution was to update a separate copy in-tree and merge the changes into the actually used root `package.json`.
This means user changes are maintained, while NodeBB dependencies are still being bumped on updates based on a git-tracked file.

It's a decent solution IMO. Feels a bit too complex, but I'm not really sure if there are simpler ways to do this without some other major sacrifices.

**Comment by julianlam** [comment]
Back to the issue at hand;

> I think dismantling dependency installation and setup (mainly mongodb connection) would be better for docker deployment. Dependencies are installed during building the docker image, and mongodb connection check is performed inside containers.

Possible, I think. If I recall correctly, the reason a database connection is required is so that we can populate the database with some default entries (the `config` object, probably). Let me check...

----

It does all of these things:

https://github.com/NodeBB/NodeBB/blob/245e5df38574af2499f223799404bc65e06ae174/src/install.js#L581-L595

So it is possible we could refactor this so that it is all run on first start, or perhaps show a dedicated welcome screen for the admin to step through (and maybe set up admin username/password) like other admin projects.

**Comment by batchfly** [comment]
@julianlam Yes, separating setup from dependency installation would be great.

But now I have a workaround to make things work. I setup a local deployment first to get the `config.json` file, then I modify the config file (mainly the mongodb host because I will be running in docker container, not localhost). At last I copied `config.json` into my docker images.

Hope this may help people with the same situation.

**Comment by mySebbe** [comment]
The concrete Docker blocker is resolved in the current repository: the Dockerfile installs `install/package.json` dependencies with `npm install --omit=dev` during image build, without a database, and leaves DB configuration/setup to runtime. `nodebb setup` still has fallback installation behavior when dependencies are absent, but containers no longer need to invoke it at build time. I suggest closing this issue for the stated deployment case. A first-start admin wizard or a CLI command split would be a separate onboarding design, not required for reproducible image builds.
